adapters/validation.py

```diff
@@ -0,0 +1,213 @@
+            node_kwargs[attr] = set()
+            if attr in attrs:
+                declarative_attrs[attr] = attrs.pop(attr)
+        cls = super().__new__(mcs, name, bases, attrs)
+        cls._declarative_attrs = declarative_attrs
+
+        for base in reversed(cls.__mro__):
+            if '_declarative_attrs' in base.__dict__:
+                for attr in mcs.DECLARATIVE_ATTRS:
+                    node_kwargs[attr].update(base._declarative_attrs.get(attr, frozenset()))
+
+        validators = []
+        if hasattr(cls, 'validators'):
+            validators.extend(cls.validators)
+        for attr in dir(cls):
+            if attr.startswith('validate') or attr.startswith('coerce'):
+                method = getattr(cls, attr)
+                if callable(method):
+                    validators.append(method)
+        node_kwargs['validators'] = validators
+
+        cls._node = ValidatorListNode(**node_kwargs)
+        return cls
+
+    def __call__(cls):
+        raise TypeError("can't instantiate {}".format(cls.__name__))
+
+    def __hash__(cls):
+        if cls is DeclarativeValidationNode:
+            return super().__hash__()
+        else:
+            return hash(cls._node)
+
+    def __eq__(cls, other):
+        if cls is other:
+            return True
+        elif cls is DeclarativeValidationNode:
+            return False
+        else:
+            return cls._node == other
+
+    def __getattr__(cls, name):
+        if name in cls.NODE_ATTRS:
+            return getattr(cls._node, name)
+        else:
+            return super().__getattr__(name)
+
+
+class DeclarativeValidationNode(object, metaclass=DeclarativeValidationNodeMeta):
+    pass
+
+
+class ValidationResult(object):
+    def __init__(self, tree, data, errors=None, failed_nodes=frozenset()):
+        self.tree = tree
+        self.data = data
+        if errors is None:
+            self.errors = []
+        else:
+            self.errors = errors
+        self.failed_nodes = frozenset(failed_nodes)
+
+    @property
+    def is_valid(self):
+        return not self.errors
+
+    def fields_with_errors(self):
+        fields = set()
+        for error in self.errors:
+            fields.add(error.field)
+        return fields
+
+
+class ValidationTree(object):
+    def __init__(self, final_node):
+        topological_order = [final_node]
+        all_nodes = {final_node}
+        nodes = {final_node}
+        while nodes:
+            node = nodes.pop()
+            new_nodes = node.dependencies - all_nodes
+            all_nodes.update(new_nodes)
+            nodes.update(new_nodes)
+            topological_order.extend(new_nodes)
+        topological_order.reverse()
+
+        nodes = {final_node}
+        for i in range(len(all_nodes) + 1):
+            node = nodes.pop()
+            nodes.update(node.dependencies)
+            if not nodes:
+                break
+        else:
+            raise CycleError('cycle in validators')
```

Comment:
I might pull this section out into a function:

```python
def check_for_cycles(root_node, node_count):
    nodes = {root_node}
    for i in range(node_count + 1):
        node = nodes.pop()
        nodes.update(node.dependencies)
        if not nodes:
            break
    else:  # We've processed each node at least once and haven't run out.
        raise CycleError('cycle in validators')

node_count = len(all_nodes)
check_for_cycles(final_node, node_count)
```

I think it makes it clearer that nothing in the scope of `__init__` we care about is mutated.